### DISCUSSION TITLE
Model-centric LLM interface

### DISCUSSION THREAD

**Discussion by andreibondarev** [patterns-ai-core/langchainrb#643] [Ideas]
## Description
When discussing the LLM interfaces with @codenamev, it's been suggested that we switch to a model-centric interface. This puts the model as the first-class citizen and the "LLM provider" as the 2nd class citizen.

The proposed interface would start with instantiating a model and then selecting the LLM provider, or the "deployment target". For example:
```ruby
Langchain::LLM::Model.new(model: "claude-3", provider: "aws_bedrock") # or "anthropic"

Langchain::LLM::Model.new(model: "command-r-plus", provider: "cohere") # or "aws_bedrock"

Langchain::LLM::Model.new(model: "llama3", provider: "groq") # or "replicate" or "hugging_face" or "ollama"
```

**Comment by codenamev** [comment]
I like this a lot 😄 Trying to think about what the interface would look like with this.  The goal being to reduce user friction to make it to the point of use faster.  Maybe something like:

```ruby
 # default provider to instantiate `Langchain::LLM::Anthropic` with env vars?
Langchain::LLM.get("claude-3").chat(...)
# or, pre-configure the LLM provider as usual
aws_bedrock_provider = Langchain::LLM::AwsBedrock.new(...)
Langchain::LLM.get("command-4-plus", provider: aws_bedrock_provider).chat(...)
```

**Comment by sandstrom** [comment]
As opposed to `llm = Langchain::LLM::OpenAI.new(api_key: '…')`?

Yes, I think that's a good idea.
